### PR TITLE
added solution method

### PR DESCRIPTION
**Pull request by tsehelnyk**

### CODE REVIEW

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 1]
```diff
@@ -3,48 +3,189 @@
...
+        this.size = 0;
+    }
+
+    private void validIndex(int index) {
+        if (index < 0 || index >= this.size) {
+            throw new ArrayIndexOutOfBoundsException("Illegal Index: " + index);
+        }
+    }
+
+    private Node<T> indexedNode(int index) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
    private Node<T> getIndexedNode(int index) {
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 2]
```diff
@@ -3,48 +3,189 @@
...
+        Node<T> node;
+        if (index < this.size / 2) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
        if (index < size / 2) {
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 3]
```diff
@@ -3,48 +3,189 @@
...
+            node = this.first;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
            node = first;
```
let's remove `this` everywhere

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 4]
```diff
@@ -3,48 +3,189 @@
...
+            node = node.prevNode;
+        }
+        return node;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == 0) {
+            Node<T> node0 = new Node<>(value);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: `node0` - rename please

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 5]
```diff
@@ -3,48 +3,189 @@
...
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, this.first.nextNode, value);
+            this.first.prevNode = newNode;
+            this.first = newNode;
+            this.size++;
+            return;
+        }
+        this.validIndex(index - 1);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: move this checking at the start of the method

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 6]
```diff
@@ -3,48 +3,189 @@
...
+            this.size++;
+            return;
+        }
+        this.add(value);
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (T i : list) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: `i` isn't very informative name of the variable

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 7]
```diff
@@ -3,48 +3,189 @@
...
-
+        this.validIndex(index);
+        if (index == 0) {
+            this.first.value = value;
+            return;
+        }
+        if (index == this.size - 1) {
+            this.last.value = value;
+            return;
+        }
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: do we need this code (95-103 lines)

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 8]
```diff
@@ -3,48 +3,189 @@
...
 
     @Override
     public T remove(int index) {
-        return null;
+        this.validIndex(index);
+        if (index == 0 && this.size == 1) {
+            Node<T> node = this.indexedNode(index);
+            this.last = null;
+            this.first = null;
+            T value = node.value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: do we need this variable?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 9]
```diff
@@ -3,48 +3,189 @@
...
+            this.size--;
+            return value;
+        }
+        if (index == 0) {
+            Node<T> nextNode = this.indexedNode(index).nextNode;
+            Node<T> node = this.indexedNode(index);
+            nextNode.prevNode = null;
+            this.first = nextNode;
+            T value = node.value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: the same

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 10]
```diff
@@ -3,48 +3,189 @@
...
+            Node<T> node = this.indexedNode(index);
+            this.last = null;
+            this.first = null;
+            T value = node.value;
+            this.size--;
+            return value;
+        }
+        if (index == 0) {
+            Node<T> nextNode = this.indexedNode(index).nextNode;
+            Node<T> node = this.indexedNode(index);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: try calling method indexedNode(index) only one time

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 11]
```diff
@@ -3,48 +3,189 @@
...
+            T value = node.value;
+            this.size--;
+            return value;
+        }
+        if (index == this.size - 1) {
+            Node<T> prevNode = this.indexedNode(index).prevNode;
+            Node<T> node = this.indexedNode(index);
+            prevNode.nextNode = null;
+            this.last = prevNode;
+            T value = node.value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: the same

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 12]
```diff
@@ -3,48 +3,189 @@
...
+            this.size--;
+            return value;
+        }
+        Node<T> prevNode = this.indexedNode(index).prevNode;
+        Node<T> nextNode = this.indexedNode(index).nextNode;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: the same

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 13]
```diff
@@ -3,48 +3,189 @@
...
+        Node<T> node = this.indexedNode(index);
+        prevNode.nextNode = nextNode;
+        nextNode.prevNode = prevNode;
+        T value = node.value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: how about removing the link on previous and next node?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 14]
```diff
@@ -3,48 +3,189 @@
...
+        this.size--;
+        return value;
     }
 
     @Override
     public T remove(T t) {
+        if (this.size > 0) {
+            Node<T> node = this.first;
+            for (int i = 0; i < size; i++) {
+                if (node.value == t || t.equals(node.value)) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: what will happen if t == null?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 15]
```diff
@@ -3,48 +3,189 @@
...
+            this.nextNode = null;
+            this.value = value;
+        }
+
+        public Node(Node<T> prevNode, Node<T> nextNode, T value) {
+            this.prevNode = prevNode;
+            this.nextNode = nextNode;
+            this.value = value;
+        }
+
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: remove redundant line

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 1 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 3 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 4 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 5 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: at the start of method we have another cheking)
they are conflict)

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 6 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 7 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done)

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 15 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 2 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 8 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: We have more than 3 lines between: 
Node<T> node = this.indexedNode(index);
and  return.
But i can't reduce the number of lines

Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 9 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 10 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 11 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 12 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: done

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 13 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: That is unnecessary. We deleted links on this Node, Garbage Collector will delete it. 
Bogdan talked about it in a lecture.)

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 14 above)
Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: i  fixed it)

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 16]
```diff
@@ -1,50 +1,174 @@
...
+        getIndexedNode(index).value = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validIndex(index);
+        if (index == 0 && size == 1) {
+            first = null;
+            Node<T> node = getIndexedNode(index);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: чи можемо ми винести цю строку після validIndex(index) і використовувати значення в кожному блоці if?

Comment (tsehelnyk) [reply src/main/java/core/basesyntax/MyLinkedList.java]: Yes, we can. Thank you!
I fixed it!)))

**Pull request closed without merging**